lib/Smr/Container/DiContainer.php

```diff
@@ -0,0 +1,112 @@
+<?php declare(strict_types=1);
+
+namespace Smr\Container;
+
+use DI\Container;
+use DI\ContainerBuilder;
+use Dotenv\Dotenv;
+use MySqlDatabase;
+use mysqli;
+use RuntimeException;
+use Smr\MySqlProperties;
+use function DI\autowire;
+
+/**
+ * Class DiContainer
+ * A wrapper around the DI\Container functionality that will allow
+ * static usage of container methods.
+ * @package Smr\Container
+ */
+class DiContainer {
+	private static DiContainer $instance;
+	private Container $container;
+
+	private function __construct() {
+		self::$instance = $this;
+		$this->container = $this->buildContainer();
+	}
+
+	private function getDefinitions(): array {
+		return [
+			/*
+			 * mysqli is a 3rd party library, and we do not have control over its constructor.
+			 * In order for PHP-DI to construct an instance of a class, each constructor argument must
+			 * be able to be constructed.
+			 * PHP-DI cannot construct mysqli by itself, because all of its arguments are primitive types.
+			 * Therefore, we need to declare a provider factory for the container to use when constructing new instances.
+			 *
+			 * The factories themselves are able to use dependency injection as well, so we can provide the MysqlProperties
+			 * typehint to make sure the container constructs and instance and provides it to the factory.
+			 */
+			mysqli::class => function (MySqlProperties $mysqlProperties): mysqli {
+				// Set the mysqli driver to raise exceptions on errors
+				if (!mysqli_report(MYSQLI_REPORT_ERROR | MYSQLI_REPORT_STRICT)) {
+					throw new RuntimeException('Failed to enable mysqli error reporting');
+				}
+				return new mysqli(
+					$mysqlProperties->getHost(),
+					$mysqlProperties->getUser(),
+					$mysqlProperties->getPassword(),
+					$mysqlProperties->getDatabaseName());
+			},
+			Dotenv::class => function (): Dotenv {
+				return Dotenv::createArrayBacked(ROOT);
+			},
+			// Explicitly name all classes that are autowired, so we can take advantage of
+			// the compiled container feature for a performance boost
+			MySqlProperties::class => autowire(),
+			MySqlDatabase::class => autowire()
+		];
+	}
+
+	private function buildContainer(): Container {
+		$builder = new ContainerBuilder();
```

Comment:
The `Container` manages object dependencies for you. That means creating them when needed, and supplying them to dependents when requested.

Please see the documentation for [compiling the container](https://php-di.org/doc/performances.html). In essence, it only happens one time per deployment. I've enabled that caching feature as sort of a pre-optimization. It exists inside the `/tmp` directory on the Docker container itself, so you don't have to worry about removing the compiled output when redeploying.

Reply:
Thanks for the information. The code comments say much the same thing, but sometimes it takes a few hits over the head for me to get it. :)

From the PHP-DI docs, I see:
> When you deploy new versions of your code to production you must delete the generated file (or the directory that contains it) to ensure that the container is re-compiled.

So I think we want to turn this off in the `smr-dev` container, since the source code is volume-mounted there to allow us to change it without rebuilding the image. (I would prefer not to overload the `PHP_DEBUG` flag for this purpose, since I often turn it off in my `smr-dev` container. Instead, maybe we can hard-wire some variable in the docker-compose configuration in the `smr` and `smr-dev` service environments.)

Reply:
Yeah good point! I will work on something